phlib/util.c

```diff
@@ -161,6 +161,226 @@ VOID PhCenterWindow(
     }
 }
 
+LONG PhGetDpi(
+    _In_ LONG Number,
+    _In_ LONG DpiValue
+    )
+{
+    return PhMultiplyDivideSigned (Number, DpiValue, USER_DEFAULT_SCREEN_DPI);
+}
+
+LONG PhGetMonitorDpi(
+    _In_ LPCRECT rect
+    )
+{
+    return PhGetDpiValue(NULL, rect);
+}
+
+LONG PhGetTaskbarDpi(
+    VOID
+    )
+{
+    APPBARDATA taskbar_rect = {0};
+
+    taskbar_rect.cbSize = sizeof (APPBARDATA);
+
+    if (SHAppBarMessage (ABM_GETTASKBARPOS, &taskbar_rect))
+        return PhGetMonitorDpi (&taskbar_rect.rc);
+
+    return PhGetDpiValue(NULL, NULL);
+}
+
+LONG PhGetWindowDpi(
+    _In_ HWND hwnd
+    )
+{
+    RECT rect;
+
+    if (!GetWindowRect (hwnd, &rect))
+        return PhGetDpiValue(hwnd, NULL);
+
+    return PhGetDpiValue(NULL, &rect);
+}
+
+LONG PhGetDpiValue(
+    _In_opt_ HWND hwnd,
+    _In_opt_ LPCRECT rect
+    )
+{
+    static PH_INITONCE initOnce = PH_INITONCE_INIT;
+    static GDFM _GetDpiForMonitor = NULL; // win81+
```

Comment:
These three GDFM, GDFW and GDFS typedefs have to be declared inline. Checkout `PhGetProcessDpiAwareness` `PhIsImmersiveProcess`, `PhGetPhysicallyInstalledSystemMemory` and `PhGetInternalWindowIcon` for some examples in `\phlib\guisup.c`